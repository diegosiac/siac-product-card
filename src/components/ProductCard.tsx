import React, { createContext } from 'react'
import { useProduct } from '../hooks/useProduct'
import { InitialValues, Product, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces'
import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export interface Props {
  // children?: React.ReactElement | React.ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: InitialValues;
  product: Product;
  style?: React.CSSProperties;
  value?: number;
  onChange?: (args: onChangeArgs) => void;
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {
  const { counter, increseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues })

  return (
    <Provider value={{
      counter,
      increseBy,
      maxCount,
      product
    }}>
        <div
          className={`${styles.productCard} ${className}`}
          style={style}
        >
            {children({
              count: counter,
              isMaxCountReached,
              maxCount: initialValues?.maxCount,
              product,
              increseBy,
              reset
            })}
        </div>
    </Provider>
  )
}

// ProductCard.Title = ProductTitle
// ProductCard.Img = ProductImage
// ProductCard.Buttons = ProductButtons
