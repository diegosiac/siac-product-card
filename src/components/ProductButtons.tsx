import React, { useCallback, useContext } from 'react'
import { ProductContext } from './ProductCard'
import styles from '../styles/styles.module.css'

export interface Props {
  className?: string;
  style?: React.CSSProperties;
}

export const ProductButtons = ({ className }: Props) => {
  const { increseBy, counter, maxCount } = useContext(ProductContext)

  const isMaxReach = useCallback(
    () => !!maxCount && maxCount === counter,
    [counter, maxCount]
  )

  return (
      <div className={`${styles.buttonsContainer} ${className}`}>
          <button onClick={() => increseBy(-1)} className={styles.buttonMinus}>-</button>
          <div className={styles.countLabel}>{counter}</div>
          <button onClick={() => increseBy(+1)} className={`${styles.buttonAdd} ${isMaxReach() && styles.disable}`}>+</button>
      </div>
  )
}
