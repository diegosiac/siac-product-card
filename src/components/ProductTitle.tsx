import React, { useContext } from 'react'
import { ProductContext } from './ProductCard'
import styles from '../styles/styles.module.css'

export interface Props {
  className?: string;
  title?: string;
  style?: React.CSSProperties;
}

export const ProductTitle = ({ title, className }: Props) => {
  const { product } = useContext(ProductContext)

  return (
      <span
        className={`${styles.productDescription} ${className}`}
      >
        {title || product.title}
      </span>
  )
}
