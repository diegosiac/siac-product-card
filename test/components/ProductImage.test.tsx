import React from 'react'
import rendered from 'react-test-renderer'
import { product2 } from '../data/products'
import { ProductCard } from '../../src/components/ProductCard'
import { ProductImage } from '../../src/components/ProductImage'

describe('ProductTitle', () => {
    test('debe de mostrar el componente correctamente con una imagen personalizada', () => {
        const wrapper = rendered.create(
            <ProductImage img='https://hola.png' className='custom-class'/>
        )

        expect(wrapper.toJSON()).toMatchSnapshot()
    })

    test('debe de mostrar el componente correctamente con la imagen del componente', () => {
        const wrapper = rendered.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot()
    });
})