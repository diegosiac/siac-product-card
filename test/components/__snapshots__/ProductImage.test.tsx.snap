// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostrar el componente correctamente con la imagen del componente 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe Mug"
    className="productImg undefined"
    src="./coffe-mug.png"
  />
</div>
`;

exports[`ProductTitle debe de mostrar el componente correctamente con una imagen personalizada 1`] = `
<img
  alt="Coffe Mug"
  className="productImg custom-class"
  src="https://hola.png"
/>
`;
